content/00.build/40.tooling/30.foundry/30.discrepancies.md

```diff
@@ -0,0 +1,125 @@
+description: Learn about differences with original foundry.
+---
+
+::callout{icon="i-heroicons-information-circle-16-solid" color="amber"}
+ZKsync foundry book is currently under active development and is going to be presented soon. Stay tuned!
+::
+
+Migrating to use ZK Foundry from Foundry involves several crucial aspects that need careful consideration.
+This document aims to comprehensively introduce these key factors, ensuring a smooth and efficient transition.
+We will cover the **fundamental** differences, potential challenges, and best practices for a successful migration.
+
+To learn more about made changes, features and limitations, please check out ZKsync Foundry [overview](/build/tooling/foundry/overview) page.
+
+::callout{icon="i-heroicons-information-circle-16-solid" color="amber"}
+Note: this guide is still work-in-progress and is going to be fullfilled in the near future.
+::
+
+## Reserved Addresses
+In ZKsync any address below `2^16` is demarcated as reserved. These addresses are often set during foundry fuzzing and can cause unwanted failures.
+To avoid these failures use `vm.assume(addr > address(65535)`
+and 65535 instances of `vm.excludeSender`(`vm.excludeSender(1)`, `vm.excludeSender(2)`, etc.)
+But this can inadvertently slow down testing and/or cause "too many rejections" failure in `proptest`, in addition to being hugely cumbersome.
+
+```sol
+  function test_BalanceIncremented() public {
+      nft.mintTo{value: 0.08 ether}(address(1));
+      uint256 slotBalance = stdstore
+          .target(address(nft))
+          .sig(nft.balanceOf.selector)
+          .with_key(address(1))
+          .find();
+
+      uint256 balanceFirstMint = uint256(
+          vm.load(address(nft), bytes32(slotBalance))
+      );
+      assertEq(balanceFirstMint, 1);
+
+      nft.mintTo{value: 0.08 ether}(address(1));
+      uint256 balanceSecondMint = uint256(
+          vm.load(address(nft), bytes32(slotBalance))
+      );
+      assertEq(balanceSecondMint, 2);
+  }
+```
+
+Introduce the fuzz config flag `no_zksync_reserved_addresses` that can be set to offset all `proptest`
+generated system reserved addresses by `65536` so they fall in the valid address range.
+
+We explicitly avoid using `prop_filter` as it can cause the same "too many rejections" failure in `proptest`,
+and instead opt to map the invalid addresses in the valid address range, so the fuzzer's *performance is practically unaffected*.
+
+A side-effect of this is that `address(0)` would no longer be the boundary value. However, since we support foundry's
+[inline test config](https://book.getfoundry.sh/reference/config/inline-test-config) the option can be selectively applied to certain tests as required.
+
+## Contract Size limitations
+In ZkEVM, the contract size must be divisible by 32 bytes (4 instructions). For example,  `bytes("mock code")` would fail the `ensure_chunkable` validation.
+Additionally, the bytecode cannot exceed `2^16` 32-byte words.
```

Comment:
This might need to be clarified with the zksolc compiler team. From what I'm aware of, there's no direct metric that can tell a user if their contract is too big. The only way to figure this out is to perform the compilation and observe the `bytecode cannot exceed 65536` error (if any).